src/test/test-Stream.ts

```diff
@@ -2084,4 +2085,312 @@ describe("Stream", () => {
 			}, 10);
 		});
 	});
+
+	describe("batch()", () => {
+		const batchResults: number[][] = [];
+		beforeEach(() => batchResults.splice(0));
+
+		function pipeWithDelay(
+			readableStream: ReadableStream<{
+				value: number;
+				workTime?: number;
+				wait?: number;
+				abort?: boolean;
+				throwError?: true;
+			}>
+		) {
+			const writableStream = new Stream<{
+				value: number;
+				workTime?: number;
+				wait?: number;
+				abort?: boolean;
+				throwError?: true;
+			}>();
+
+			writableStream.aborted().catch((e) => readableStream.abort(e));
+			readableStream.aborted().catch((e) => writableStream.abort(e));
+
+			readableStream
+				.forEach(async (item) => {
+					const { wait } = item;
+
+					if (wait !== undefined) {
+						await delay(wait);
+						writableStream.write(item);
+					} else {
+						writableStream.write(item);
+					}
+				})
+				.then(() => writableStream.end())
+				.catch(() => writableStream.end());
+
+			return writableStream;
+		}
+
+		async function resolveBatchToAsyncValues(
```

Comment:
This is interesting. I tried doing something like this but found that I had to think very carefully about how to avoid deadlocks when `await`ing things. It didn't seem easier to follow to me.

I did go through this function however and simplified it a lot.

Reply:
One interesting technique that may help prevent such kind of hard-to-sync stuff for these cases is that you can start two 'tasks' in parallel: one for the source side, and one for the sink side. Just put each in their own async function, start both, then `await Promise.all([sourceTask, sinkTask]);`

Reply:
BTW, after reviewing your updated tests just now, I think you've done a great job, so no need to change anything anymore.

Reply:
Thanks!